pulsar-client/src/main/java/org/apache/pulsar/client/impl/TransactionMetaStoreHandler.java

```diff
@@ -335,41 +402,77 @@ void handleEndTxnResponse(CommandEndTxnResponse response) {
+                timer.newTimeout(timeout -> {
+                    long requestId = client.newRequestId();
+                    TxnID txnID = op.txnID;
+                    TxnAction action = op.action;
+                    BaseCommand cmd = Commands.newEndTxn(requestId, txnID.getLeastSigBits(),
+                            txnID.getMostSigBits(), action);
+                    OpForEndTxnCallBack opNew = OpForEndTxnCallBack
+                            .create(Commands.serializeWithSize(cmd), op.callback, txnID, action);
+                    op.cmd.release();
+                    op.recycle();
+                    tryExecuteCommandAgain(opNew, requestId);
+                }, op.backoff.next(), TimeUnit.MILLISECONDS);
+                return;
+            } else {
+                LOG.error("Got end txn response for request {} error {}", response.getRequestId(), response.getError());
+            }
         }
-
         onResponse(op);
     }
 
+    private <T> void tryExecuteCommandAgain(OpBase<T> op, long requestId) {
+        if (cnx() == null) {
+            timer.newTimeout(timeout ->
+                    tryExecuteCommandAgain(op, requestId), op.backoff.next(), TimeUnit.MILLISECONDS);
+            return;
+        }
+        pendingRequests.put(requestId, op);
+        timeoutQueue.add(new RequestTime(System.currentTimeMillis(), requestId));
+        cnx().ctx().writeAndFlush(op.cmd, cnx().ctx().voidPromise());
+    }
+
     private void handleTransactionFailOp(ServerError error, String message, OpBase<?> op) {
         if (error == ServerError.TransactionCoordinatorNotFound && getState() != State.Connecting) {
             connectionHandler.reconnectLater(new TransactionCoordinatorClientException
                     .CoordinatorNotFoundException(message));
         }
 
-        if (op != null) {
+        if (op != null && error != ServerError.TransactionCoordinatorNotFound) {
             op.callback.completeExceptionally(getExceptionByServerError(error, message));
         }
     }
 
     private static abstract class OpBase<T> {
         protected ByteBuf cmd;
         protected CompletableFuture<T> callback;
+        protected Backoff backoff = new Backoff(100, TimeUnit.MILLISECONDS, 3, TimeUnit.SECONDS, 10,
+                TimeUnit.SECONDS);
 
         abstract void recycle();
     }
 
-    private static class OpForTxnIdCallBack extends OpBase<TxnID> {
+    private static class OpForNewTxnCallBack extends OpBase<TxnID> {
 
-        static OpForTxnIdCallBack create(ByteBuf cmd, CompletableFuture<TxnID> callback) {
-            OpForTxnIdCallBack op = RECYCLER.get();
+        protected long timeout;
+        protected TimeUnit unit;
+        static OpForNewTxnCallBack create(ByteBuf cmd, CompletableFuture<TxnID> callback, long timeout, TimeUnit unit) {
+            OpForNewTxnCallBack op = RECYCLER.get();
             op.callback = callback;
             op.cmd = cmd;
+            op.timeout = timeout;
+            op.unit = unit;
```

Comment:
every field need to be set initial value like `org.apache.bookkeeper.mledger.impl.OpReadEntry.recycle()`